control-plane/config-entries/controllersv2/mesh_gateway_controller.go

```diff
@@ -69,11 +72,89 @@ func (r *MeshGatewayController) SetupWithManager(mgr ctrl.Manager) error {
 }
 
 func (r *MeshGatewayController) onCreateUpdate(ctx context.Context, req ctrl.Request, resource *meshv2beta1.MeshGateway) error {
-	// TODO NET-6392 NET-6393 NET-6394 NET-6395
-	return errors.New("onCreateUpdate not implemented")
+	builder := gateways.NewMeshGatewayBuilder(resource)
+
+	upsertOp := func(ctx context.Context, _, object client.Object) error {
+		_, err := controllerutil.CreateOrUpdate(ctx, r.Client, object, func() error { return nil })
+		return err
+	}
+
+	err := r.opIfNewOrOwned(ctx, resource, &corev1.ServiceAccount{}, builder.ServiceAccount(), upsertOp)
+	if err != nil {
+		return err
+	}
+
+	// TODO NET-6392 NET-6393 NET-6395
+
+	return nil
 }
 
 func (r *MeshGatewayController) onDelete(ctx context.Context, req ctrl.Request, resource *meshv2beta1.MeshGateway) error {
-	// TODO NET-6392 NET-6393 NET-6394 NET-6395
-	return errors.New("onDelete not implemented")
+	builder := gateways.NewMeshGatewayBuilder(resource)
+
+	deleteOp := func(ctx context.Context, _, object client.Object) error {
```

Comment:
I'm thinking we can just let the garbage collector clean these up? By setting the owner reference field, if the owner is deleted, these will actually get cleaned up automatically. If we're cool with just letting the GC do its work, then the only thing we have to worry about on deletion is cleaning up any side-effects that we may have done in our controller upsert path.

Reply:
My assumption was that there's a reason we do explicit cleanup in our v1 implementation [here](https://github.com/hashicorp/consul-k8s/blob/671675d49fbca59272fbbd54becd6366620704a5/control-plane/api-gateway/gatekeeper/gatekeeper.go#L60-L86), but I don't know what that might have been. Any ideas?

Reply:
I've dropped the explicit deletes in favor of allowing the k8s garbage collector to do its thing.
If we discover a reason we need to delete explicitly like we do in the `gatekeeper` package today, then we can add the explicit deletion at a later date.